### PR TITLE
feat(backend): update to mysql8. Fixes #5893

### LINKED ISSUE #5893: [backend] column name same with reserved keyword of mysql 8.0
### Environment

*  How did you deploy Kubeflow Pipelines (KFP)?
* create table in mysql and deploy the KFP services

*  KFP version: 
* 1.5

*  KFP SDK version: 

### Steps to reproduce
When creating table in mysql 8.0, it would report fail because there are columns name `Description` in tables `experiments`, `jobs`, `pipelines`, `run_details`, which is the reserved keyword of mysql

### Expected result
Creating tables without error.

Shall we use `Descriptions` as the columns' name?

### Materials and Reference

---

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

cc @Jeffwan

### PR DESCRIPTION
**Pull request by NikeNano**
**Description of your changes:**
Update to mysql8, fixes https://github.com/kubeflow/pipelines/issues/5893

[UPDATED]

In order to check for issues upgrading mysql(from 5.7) to version 8.0.25 the [upgrade checker](https://dev.mysql.com/doc/mysql-shell/8.0/en/mysql-shell-utilities-upgrade.html) where used. 

```
kubectl port-forward -n kubeflow svc/mysql 3306:3306
mysqlsh -u root -h 127.0.0.1
```

```
util.checkForServerUpgrade()
```
which gives

```
The MySQL server at 127.0.0.1:3306, version 5.7.33 - MySQL Community Server
(GPL), will now be checked for compatibility issues for upgrade to MySQL
8.0.25...

1) Usage of old temporal type
  No issues found

2) Usage of db objects with names conflicting with new reserved keywords
  No issues found

3) Usage of utf8mb3 charset
  No issues found

4) Table names in the mysql schema conflicting with new tables in 8.0
  No issues found

5) Partitioned tables using engines with non native partitioning
  No issues found

6) Foreign key constraint names longer than 64 characters
  No issues found

7) Usage of obsolete MAXDB sql_mode flag
  No issues found

8) Usage of obsolete sql_mode flags
  Notice: The following DB objects have obsolete options persisted for
    sql_mode, which will be cleared during upgrade to 8.0.
  More information:
    https://dev.mysql.com/doc/refman/8.0/en/mysql-nutshell.html#mysql-nutshell-removals

  global system variable sql_mode - defined using obsolete NO_AUTO_CREATE_USER
    option

9) ENUM/SET column definitions containing elements longer than 255 characters
  No issues found

10) Usage of partitioned tables in shared tablespaces
  No issues found

11) Circular directory references in tablespace data file paths
  No issues found

12) Usage of removed functions
  No issues found

13) Usage of removed GROUP BY ASC/DESC syntax
  No issues found

14) Removed system variables for error logging to the system log configuration
  To run this check requires full path to MySQL server configuration file to be specified at 'configPath' key of options dictionary
  More information:
    https://dev.mysql.com/doc/relnotes/mysql/8.0/en/news-8-0-13.html#mysqld-8-0-13-logging

15) Removed system variables
  To run this check requires full path to MySQL server configuration file to be specified at 'configPath' key of options dictionary
  More information:
    https://dev.mysql.com/doc/refman/8.0/en/added-deprecated-removed.html#optvars-removed

16) System variables with new default values
  To run this check requires full path to MySQL server configuration file to be specified at 'configPath' key of options dictionary
  More information:
    https://mysqlserverteam.com/new-defaults-in-mysql-8-0/

17) Zero Date, Datetime, and Timestamp values
  No issues found

18) Schema inconsistencies resulting from file removal or corruption
  No issues found

19) Tables recognized by InnoDB that belong to a different engine
  No issues found

20) Issues reported by 'check table x for upgrade' command
  No issues found

21) New default authentication plugin considerations
  Warning: The new default authentication plugin 'caching_sha2_password' offers
    more secure password hashing than previously used 'mysql_native_password'
    (and consequent improved client connection authentication). However, it also
    has compatibility implications that may affect existing MySQL installations.
    If your MySQL installation must serve pre-8.0 clients and you encounter
    compatibility issues after upgrading, the simplest way to address those
    issues is to reconfigure the server to revert to the previous default
    authentication plugin (mysql_native_password). For example, use these lines
    in the server option file:

    [mysqld]
    default_authentication_plugin=mysql_native_password

    However, the setting should be viewed as temporary, not as a long term or
    permanent solution, because it causes new accounts created with the setting
    in effect to forego the improved authentication security.
    If you are using replication please take time to understand how the
    authentication plugin changes may impact you.
  More information:
    https://dev.mysql.com/doc/refman/8.0/en/upgrading-from-previous-series.html#upgrade-caching-sha2-password-compatibility-issues
    https://dev.mysql.com/doc/refman/8.0/en/upgrading-from-previous-series.html#upgrade-caching-sha2-password-replication

Errors:   0
Warnings: 1
Notices:  1

No fatal errors were found that would prevent an upgrade, but some potential issues were detected. Please ensure that the reported issues are not significant before upgrading.
```

Due to the upgrade in mysql: 
```
Because the data dictionary provides information about database objects, the server no longer checks directory names in the data directory to find databases. Consequently, the --ignore-db-dir option is extraneous and has been removed. To handle this, remove any instances of --ignore-db-dir from your startup configuration
```

```
 --ignore-db-di
```

 can be dropped.

Also need to update: 

- .cloudbuild.yaml
- .release.cloudbuild.yaml

Could you help out with how to update the images and how to update the files above @Bobgy ?

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: manifests/kustomize/third-party/mysql/base/mysql-deployment.yaml [hunk 1]
```diff
@@ -20,13 +20,12 @@ spec:
...
       # Ext4, Btrfs etc. volumes root directories have a lost+found directory that should not be treated as a database.
       - args:
-        - --ignore-db-dir=lost+found
         - --datadir
         - /var/lib/mysql
         env:
         - name: MYSQL_ALLOW_EMPTY_PASSWORD
           value: "true"
-        image: gcr.io/ml-pipeline/mysql:5.7
+        image: docker.io/mysql:8.0
```
Comment (Bobgy) [inline manifests/kustomize/third-party/mysql/base/mysql-deployment.yaml:28]: Can you post a link to its Dockerfile?
We need to make sure it is compliant with licenses of Its dependencies and transitive dependencies.

File: manifests/kustomize/third-party/mysql/base/mysql-deployment.yaml [hunk 2]
```diff
@@ -20,13 +20,12 @@ spec:
       # https://dev.mysql.com/doc/refman/5.7/en/server-options.html#option_mysqld_ignore-db-dir
       # Ext4, Btrfs etc. volumes root directories have a lost+found directory that should not be treated as a database.
       - args:
-        - --ignore-db-dir=lost+found
         - --datadir
         - /var/lib/mysql
         env:
         - name: MYSQL_ALLOW_EMPTY_PASSWORD
           value: "true"
-        image: gcr.io/ml-pipeline/mysql:5.7
```
Comment (Bobgy) [inline manifests/kustomize/third-party/mysql/base/mysql-deployment.yaml:29]: Does MySQL 5.7 to 8 guarantee backwards compatibility? Anything affects us that changed?

**Review by Bobgy — COMMENTED** [review]
Thank you for keeping dependencies up to date!
Left some detailed comments

File: manifests/kustomize/third-party/mysql/base/mysql-deployment.yaml (hunk 2 above)
Comment (NikeNano) [reply manifests/kustomize/third-party/mysql/base/mysql-deployment.yaml:29]: Will update with documentation on this.

Comment (NikeNano) [reply manifests/kustomize/third-party/mysql/base/mysql-deployment.yaml:29]: Updated in the PR description

File: manifests/kustomize/third-party/mysql/base/mysql-deployment.yaml (hunk 1 above)
Comment (NikeNano) [reply manifests/kustomize/third-party/mysql/base/mysql-deployment.yaml:28]: missed to send this yeasterday. [dockerhub](https://hub.docker.com/layers/mysql/library/mysql/8.0.25/images/sha256-c1afde725e2cfb627136a299b4d4bd35ae30a31fae1297dd2b3c3c951d9c7240?context=explore), github: https://github.com/docker-library/mysql/blob/a09a716a88ce34a17e16df3f8b091066d90e6f34/8.0/Dockerfile.debian

**Comment by NikeNano** [comment]
from: kubeflow-pipeline-mkp-test

```
Step #32 - "tagForHosted": Error response from daemon: No such image: gcr.io/ml-pipeline/mysql:8.0
```

What is needed to update/create the needed images @Bobgy?

**Comment by Bobgy** [comment]
We need to push mysql8 image to gcr.io/ml-pipeline/mysql:8.0 after confirming license compliance https://github.com/kubeflow/pipelines/pull/5946#discussion_r661026722

**Comment by Bobgy** [comment]
Any updates on confirming license compliance?
Or I can take a look too

**Comment by NikeNano** [comment]
> Any updates on confirming license compliance?
> Or I can take a look too

Ohh sorry my bad @Bobgy , I though you where going to look into it. I can take a look if you can bring me up to speed at what we are looking for and what it means that the license is compliant?

**Comment by NikeNano** [comment]
> > Any updates on confirming license compliance?
> > Or I can take a look too
> 
> Ohh sorry my bad @Bobgy , I though you where going to look into it. I can take a look if you can bring me up to speed at what we are looking for and what it means that the license is compliant?

@Boby what should I look into in relation to licences? Do we have some documentation for this

**Comment by Bobgy** [comment]
@NikeNano sorry, didn't have time to check right now, was focused on blockers for 1.7.0 release.

Basically, you would need to confirm whether all the binaires and code in the container complies to their license.

For pkg downloaded by apk/apt, you can skip them, because they are already compliant.

For any built binary, we need to ensure all transitive dependencies have their license file copied in the image, and for LGPL, GPL, MPL etc licenses, we need to ensure their source code is in the image.

And no code brought in should have incompatible licenses with KFP. No AGPL.

**Comment by Bobgy** [comment]
If you are not sure, identifying what needs to be checked is also helpful!

**Comment by NikeNano** [comment]
> Good to know, then let me add MySQL to KFP gcr.

Have you had the opportunity to add the image @Bobgy ?

**Comment by NikeNano** [comment]
> Good to know, then let me add MySQL to KFP gcr.

Is this updated @Bobgy ?

**Comment by NikeNano** [comment]
@chensun or @capri-xiyue do any of you have the possibility to update the images required? Thank you

**Comment by zijianjoy** [comment]
Hello @NikeNano , sorry for the delay. I have pushed the mysql 8.0 image to Google container registry: https://console.cloud.google.com/gcr/images/ml-pipeline/global/mysql@sha256:238cf050a7270dd6940602e70f1e5a11eeaf4e02035f445b7f613ff5e0641f7d/details.

The tag is: gcr.io/ml-pipeline/mysql:8.0

**Comment by NikeNano** [comment]
> Hello @NikeNano , sorry for the delay. I have pushed the mysql 8.0 image to Google container registry: https://console.cloud.google.com/gcr/images/ml-pipeline/global/mysql@sha256:238cf050a7270dd6940602e70f1e5a11eeaf4e02035f445b7f613ff5e0641f7d/details.
> 
> The tag is: gcr.io/ml-pipeline/mysql:8.0

Awesome, I will make the update during the weekend then.

**Comment by zijianjoy** [comment]
Fixing kubeflow-pipeline-mkp-snapshot-test  with https://github.com/kubeflow/pipelines/pull/7166

**Comment by GerardoGR** [comment]
Is there any open topic for this issue?. It would be great to have this merged to avoid [really high memory usage in mysql](https://github.com/docker-library/mysql/issues/579) in setups where the `nofile` is set too high or infinity (now a days a default for containerd: https://github.com/containerd/containerd/pull/4475).

Also worth mentioning, AFAIK the [ml-metadata doesn't support caching_sha2_password authentication](https://github.com/google/ml-metadata/issues/128) ([the default one for mysql 8](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_default_authentication_plugin)), then to have this running in the `mysql-deployment` we need to have these args:

```yaml
      - args:
        # NOTE: use of "mysql_native_password" is not recommended: https://dev.mysql.com/doc/refman/8.0/en/upgrading-from-previous-series.html#upgrade-caching-sha2-password
        # (this is just an example, not intended to be a production configuration)
        - --default-authentication-plugin=mysql_native_password
        - --datadir
        - /var/lib/mysql
```

**Comment by thesuperzapper** [comment]
@zijianjoy I am not sure about this PR in particular, but what is the status on supporting MySQL 8.0 in Kubeflow Pipelines backend?

**Comment by zijianjoy** [comment]
We are already supporting mysql 8 as in https://github.com/kubeflow/pipelines/commit/65ee01a436758e0f7ca4741e517452c02182ec03

**Comment by thesuperzapper** [comment]
@zijianjoy I will close this then, as https://github.com/kubeflow/pipelines/pull/8351 has done what this PR was trying to do.

/close

**Pull request closed without merging**